std/format.d

```diff
@@ -2628,55 +2632,118 @@ if (is(FloatingPointTypeOf!T) && !is(T == enum) && !hasToString!(T, Char))
     const nan = isNaN(tval);
     const inf = isInfinity(tval);
 
-    if (nan || inf)
+    static if (is(T == float) || is(T == double) || (is(T == real) && T.mant_dig == double.mant_dig))
     {
-        const sb = signbit(tval);
-        const up = isUpper(spec);
-        string ns = nanInfStr(f, nan, inf, sb, up);
-        FormatSpec!Char co;
-        co.spec = 's';
-        co.width = f.width;
-        co.flDash = f.flDash;
-        formatValue(w, ns, co);
-        return;
+        size_t len;
+        char[] buf;
+
+        if (fs.spec=='a' || fs.spec=='A')
+        {
+            buf = printFloat(val, fs, FloatingPointControl.rounding).dup;
+            len = buf.length;
+        }
+        else
+        {
+            if (nan || inf)
+            {
+                const sb = signbit(tval);
+                const up = isUpper(spec);
+                string ns = nanInfStr(f, nan, inf, sb, up);
+                FormatSpec!Char co;
+                co.spec = 's';
+                co.width = f.width;
+                co.flDash = f.flDash;
+                formatValue(w, ns, co);
+                return;
+            }
+
+            char[1 /*%*/ + 5 /*flags*/ + 3 /*width.prec*/ + 2 /*format*/
+                 + 1 /*\0*/] sprintfSpec = void;
+            sprintfSpec[0] = '%';
+            uint i = 1;
+            if (fs.flDash) sprintfSpec[i++] = '-';
+            if (fs.flPlus) sprintfSpec[i++] = '+';
+            if (fs.flZero) sprintfSpec[i++] = '0';
+            if (fs.flSpace) sprintfSpec[i++] = ' ';
+            if (fs.flHash) sprintfSpec[i++] = '#';
+            sprintfSpec[i .. i + 3] = "*.*";
+            i += 3;
+            if (is(Unqual!(typeof(val)) == real)) sprintfSpec[i++] = 'L';
+            sprintfSpec[i++] = spec2;
+            sprintfSpec[i] = 0;
+            //printf("format: '%s'; geeba: %g\n", sprintfSpec.ptr, val);
+            char[512] buf_tmp = void;
+
+            //writefln("'%s'", sprintfSpec[0 .. i]);
+
+            immutable n = ()@trusted{
+                import core.stdc.stdio : snprintf;
+                return snprintf(buf_tmp.ptr, buf_tmp.length,
+                                sprintfSpec.ptr,
+                                fs.width,
+                                // negative precision is same as no precision specified
+                                fs.precision == fs.UNSPECIFIED ? -1 : fs.precision,
+                                tval);
+            }();
+
+            enforceFmt(n >= 0,
+                       "floating point formatting failure");
+
+            len = min(n, buf_tmp.length-1);
+            buf = buf_tmp;
```

Comment:
Not that it seems used later but assigning to a stack variable that is about to go out of scope is a bad idea. Also this is probably the reason for your `@safe`ty problem.

Reply:
It's a workaround until snprintf is removed completely. After that this is not needed anymore. I'll see if I find a way to do this better.